### ISSUE TITLE
Missing rules result in a large report range

### ISSUE THREAD

**Issue by not-an-aardvark** [AtomLinter/linter-eslint#908]
## Issue Type

Bug, sort of

## Issue Description

When using ESLint 4.0.0-alpha, I added a rule that does not exist to my config file. As expected, ESLint reported an error because the rule was missing. However, linter-eslint displayed the error like this:

![](https://i.gyazo.com/8ec7e9428619efcb112cba507dd21d6e.png)

In other words, it highlighted the entire file rather than a single line, which is what it did previously.

The issue is caused by [this change](http://eslint.org/docs/4.0.0/user-guide/migrating-to-4.0.0#report-locations) in ESLint 4.0. If a rule reports a node, the reported problem now has an end range. In this case, the entire AST was reported, so there are a lot of highlighted lines.

It might be worth limiting the number of lines highlighted in cases like this.

## Bug Checklist
- Restart Atom
- Verify the `eslint` CLI gives the proper result, while `linter-eslint` does not
- Paste the output of the `Linter Eslint: Debug` command from the Command Palette below

```
Atom version: 1.16.0
linter-eslint version: 8.1.7
ESLint version: 4.0.0-alpha.1
Hours since last Atom restart: 0.1
Platform: darwin
Using local project ESLint from: /path/to/eslint/node_modules/eslint
Current file's scopes: [
  "source.js",
  "meta.function.js",
  "meta.parameters.js",
  "punctuation.definition.parameters.begin.bracket.round.js"
]
linter-eslint configuration: {
  "disableFSCache": true,
  "eslintRulesDir": "lib/internal-rules",
  "rulesToSilenceWhileTyping": [
    "no-trailing-spaces"
  ],
  "lintHtmlFiles": false,
  "useGlobalEslint": false,
  "showRuleIdInMessage": true,
  "disableWhenNoEslintConfig": true,
  "eslintrcPath": "",
  "globalNodePath": "",
  "advancedLocalNodeModules": "",
  "disableEslintIgnore": false,
  "fixOnSave": false,
  "scopes": [
    "source.js",
    "source.jsx",
    "source.js.jsx",
    "source.babel",
    "source.js-semantic"
  ],
  "rulesToDisableWhileFixing": []
}
```

**Comment by Arcanemagus** [comment]
Known issue, thanks for filing it to make sure it gets handled though!

**Comment by Arcanemagus** [comment]
@not-an-aardvark Will these "entire file" messages _always_ have a `nodeType` of `Program`?

**Comment by not-an-aardvark** [comment]
The particular error message for missing rules comes from [here](https://github.com/eslint/eslint/blob/3d020b9083de105ce8adc67bc3efae0aac01e03f/lib/linter.js#L522-L524), so it currently has a `nodeType` of `Program`. (I don't want to guarantee that the missing rule message will *always* have that node type, because it's possible we would change the report location or node type, e.g. to be less intrusive in cases like this).

In general, it's possible that rules will report a "large" range as a result of this change (which might be a `Program` node, or it might be some other "large" node, e.g. a `FunctionExpression` node which spans most of the file). The idea is that integrations could choose to behave differently for messages with very large report ranges, rather than filtering for "missing rule" messages in particular.

**Comment by Arcanemagus** [comment]
Hmmm, I'm fine with filtering down messages that highlight _literally_ the entire file, but I'm not sure I'd want to filter out ones that "just" highlight a large section of it.

I think with the current solution I can actually remove the nodeType check, so I'll do that.

**Comment by Arcanemagus** [comment]
@IanVS Do you think we should filter "very large" ranges? And if so, what would be the threshold?

**Comment by IanVS** [comment]
Oof, that gets tricky.  If we decide on some number of lines, I would expect users to become confused by the seemingly inconsistent behavior.  I think a good starting point is to make sure the entire file is not being highlighted, but personally I wouldn't want to set a threshold to truncate/remove selection for "large ranges."

**Comment by Arcanemagus** [comment]
That's how I feel on it as well, #969 _should_ handle this then.

**Comment by not-an-aardvark** [comment]
It's worth noting that the missing rule messages will have a range spanning the entire `Program` node (between the first token and the last token of the file), which is not necessarily the same as the entire file. For example, if there is whitespace or a comment before the start of the program, the report range will start after that.

---

As a separate note, I'm wondering if we should change the behavior of the missing rule messages within ESLint to just report at `{line: 1, column: 1}` rather than reporting the entire file. Reporting the entire file was actually a side-effect of the change (the message is reported at the `Program` node, which was previously about the same as being reported at the top of the file).

**Comment by Arcanemagus** [comment]
> For example, if there is whitespace or a comment before the start of the program, the report range will start after that.

Ah, interesting. Would comments before the first code be highlighted or would it only start at the first bit of code? Currently #969 is only stripping pure whitespace from the end, I didn't test comments at the end or whitespace/comments at the start.

----

> As a separate note, I'm wondering if we should change the behavior of the missing rule messages within ESLint to just report at `{line: 1, column: 1}`

That's essentially what #969 does.

**Comment by not-an-aardvark** [comment]
> Would comments before the first code be highlighted or would it only start at the first bit of code?

I think it would just start at the first bit of code.

> That's essentially what #969 does.

Right, I meant that the message itself should be changed upstream in ESLint, rather than adding a special case to display the message in linter-eslint.

**Comment by Arcanemagus** [comment]
> I think it would just start at the first bit of code.

That's what I was afraid of. Determining where the first "non-comment" line is sounds a bit out of scope for what is likely to be a quickly fixed bug in configuration. I'm beginning to wonder if we shouldn't even bother with special case handling of these issues here.

> I meant that the message itself should be changed upstream in ESLint

Sorry, meant that as just a "yea, that's what this is doing anyway". It sounds like a great idea to me!

**Comment by not-an-aardvark** [comment]
Created https://github.com/eslint/eslint/pull/8998.

As a separate note, it's possible that some more rules will start having large report ranges soon (https://github.com/eslint/eslint/issues/8980)

**Comment by Arcanemagus** [comment]
I'm going to mark this as `wontfix` here, "fixing" it correctly would involve a large amount of heuristic logic to determine what a "large report" is, and as rules reporting such large ranges are a rare occurrence in the first place I don't think this is a large enough issue to warrant the large amount of work that would be required for that.
